php/ext/google/protobuf/message.c

```diff
@@ -342,3 +343,322 @@ PHP_METHOD(Message, whichOneof) {
+                                                NULL PHP_PROTO_TSRMLS_CC);
+#else
+  zval* value = message_handlers->read_property(getThis(), &member, BP_VAR_R,
+                                                NULL, NULL PHP_PROTO_TSRMLS_CC);
+#endif
+  EG(scope) = old_scope;
+
+#if PHP_MAJOR_VERSION < 7
+  ZVAL_ZVAL(return_value, value, 1, 0);
+#else
+  ZVAL_COPY(return_value, value);
+#endif
+}
+
+PHP_METHOD(Any, setValue) {
+  char *value = NULL;
+  PHP_PROTO_SIZE value_len;
+
+  if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "s", &value,
+                            &value_len) == FAILURE) {
+    return;
+  }
+
+  zval member;
+  zval value_to_set;
+  PHP_PROTO_ZVAL_STRING(&member, "value", 1);
+  PHP_PROTO_ZVAL_STRINGL(&value_to_set, value, value_len, 1);
+
+  zend_class_entry* old_scope = EG(scope);
+  EG(scope) = any_type;
+  message_handlers->write_property(getThis(), &member, &value_to_set,
+                                   NULL PHP_PROTO_TSRMLS_CC);
+  EG(scope) = old_scope;
+
+#if PHP_MAJOR_VERSION < 7
+  ZVAL_ZVAL(return_value, getThis(), 1, 0);
+#else
+  ZVAL_COPY(return_value, getThis());
+#endif
+}
+
+PHP_METHOD(Any, unpack) {
+  // Get type url.
+  zval type_url_member;
+  PHP_PROTO_ZVAL_STRING(&type_url_member, "type_url", 1);
+  zend_class_entry* old_scope = EG(scope);
+  EG(scope) = any_type;
+#if PHP_MAJOR_VERSION < 7
+  zval* type_url_php = message_handlers->read_property(
+      getThis(), &type_url_member, BP_VAR_R, NULL PHP_PROTO_TSRMLS_CC);
+#else
+  zval* type_url_php = message_handlers->read_property(
+      getThis(), &type_url_member, BP_VAR_R, NULL, NULL PHP_PROTO_TSRMLS_CC);
+#endif
+  EG(scope) = old_scope;
+
+  // Get fully-qualified name from type url.
+  const char* url_prefix = "type.googleapis.com/";
+  size_t url_prefix_len = strlen(url_prefix);
+  const char* type_url = Z_STRVAL_P(type_url_php);
+  size_t type_url_len = Z_STRLEN_P(type_url_php);
+
+  if (url_prefix_len > type_url_len ||
+      strncmp(url_prefix, type_url, url_prefix_len) != 0) {
+    zend_throw_exception(
+        NULL, "Type url needs to be type.googleapis.com/fully-qulified",
+        0 TSRMLS_CC);
+    return;
+  }
+
+  const char* fully_qualified_name = type_url + url_prefix_len;
+  PHP_PROTO_HASHTABLE_VALUE desc_php = get_proto_obj(fully_qualified_name);
+  if (desc_php == NULL) {
+    zend_throw_exception(
+        NULL, "Specified message in any hasn't been added to descriptor pool",
+        0 TSRMLS_CC);
+    return;
+  }
+  Descriptor* desc = UNBOX_HASHTABLE_VALUE(Descriptor, desc_php);
+  zend_class_entry* klass = desc->klass;
+  ZVAL_OBJ(return_value, klass->create_object(klass TSRMLS_CC));
+  MessageHeader* msg = UNBOX(MessageHeader, return_value);
+  custom_data_init(klass, msg PHP_PROTO_TSRMLS_CC);
+
+  // Get value.
+  zval value_member;
+  PHP_PROTO_ZVAL_STRING(&value_member, "value", 1);
+  old_scope = EG(scope);
+  EG(scope) = any_type;
+#if PHP_MAJOR_VERSION < 7
```

Comment:
This pattern happens quite a lot. Perhaps worth it to put a macro in `protobuf.h`?